### PR TITLE
[WIP] Correctly set terminated pods phase, reason, message, and statuses

### LINKED ISSUE #105358: Pods go into error state after kubelet-initiated eviction
#### What happened:

When a pod is evicted under resource pressure, it goes into an error state.

#### What you expected to happen:

The pod terminates normally.

#### How to reproduce it (as minimally and precisely as possible):

Note that the 130GB file is based on **90% of total disk space - current utilization** to hit the default eviction threshold.

```bash
ehashman@fedora:~/src/k8s$ cat ~/tmp/disk-pressure.yml 
kind: Pod
apiVersion: v1
metadata:
  name: evictme
spec:
  terminationGracePeriodSeconds: 10
  containers:
  - name: busybox
    image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
    command: ["sh", "-c", "fallocate -l 130G file; sleep 100000"]

ehashman@fedora:~/src/k8s$ cluster/kubectl.sh get pod
No resources found in default namespace.

ehashman@fedora:~/src/k8s$ cluster/kubectl.sh apply -f ~/tmp/disk-pressure.yml 
pod/evictme created

ehashman@fedora:~/src/k8s$ cluster/kubectl.sh get pod
NAME      READY   STATUS    RESTARTS   AGE
evictme   1/1     Running   0          3s

ehashman@fedora:~/src/k8s$ cluster/kubectl.sh get pod
NAME      READY   STATUS    RESTARTS   AGE
evictme   1/1     Evicted   0          10s

ehashman@fedora:~/src/k8s$ cluster/kubectl.sh get pod
NAME      READY   STATUS   RESTARTS   AGE
evictme   0/1     Error    0          21s
```

Full pod spec after going into error state:

<details>

```
ehashman@fedora:~/src/k8s$ cluster/kubectl.sh get pod -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"evictme","namespace":"default"},"spec":{"containers":[{"command":["sh","-c","fallocate -l 130G file; sleep 10000"],"image":"k8s.gcr.io/e2e-test-images/nginx:1.14-2","name":"busybox"}],"terminationGracePeriodSeconds":10}}
    creationTimestamp: "2021-10-01T19:10:20Z"
    name: evictme
    namespace: default
    resourceVersion: "519"
    uid: e9e27b78-d722-4c2c-96f3-7a22877d75bb
  spec:
    containers:
    - command:
      - sh
      - -c
      - fallocate -l 130G file; sleep 10000
      image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imagePullPolicy: IfNotPresent
      name: busybox
      resources: {}
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: kube-api-access-dctrp
        readOnly: true
    dnsPolicy: ClusterFirst
    enableServiceLinks: true
    nodeName: 127.0.0.1
    preemptionPolicy: PreemptLowerPriority
    priority: 0
    restartPolicy: Always
    schedulerName: default-scheduler
    securityContext: {}
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 10
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    volumes:
    - name: kube-api-access-dctrp
      projected:
        defaultMode: 420
        sources:
        - serviceAccountToken:
            expirationSeconds: 3607
            path: token
        - configMap:
            items:
            - key: ca.crt
              path: ca.crt
            name: kube-root-ca.crt
        - downwardAPI:
            items:
            - fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
              path: namespace
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T19:10:59Z"
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T19:11:13Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T19:11:13Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T19:10:59Z"
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: docker://ea630f3b1b7e83763df0693d0e617b475433486462abd0ec01611f146c883516
      image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imageID: docker-pullable://k8s.gcr.io/e2e-test-images/nginx@sha256:13616070e3f29de4417eee434a8ef472221c9e51b3d037b5a6b46cef08eb7443
      lastState: {}
      name: busybox
      ready: false
      restartCount: 0
      started: false
      state:
        terminated:
          containerID: docker://ea630f3b1b7e83763df0693d0e617b475433486462abd0ec01611f146c883516
          exitCode: 137
          finishedAt: "2021-10-01T19:11:13Z"
          reason: Error
          startedAt: "2021-10-01T19:11:00Z"
    hostIP: 127.0.0.1
    phase: Running
    podIP: 172.17.0.3
    podIPs:
    - ip: 172.17.0.3
    qosClass: BestEffort
    startTime: "2021-10-01T19:10:59Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
</details>

#### Anything else we need to know?:

See also https://bugzilla.redhat.com/show_bug.cgi?id=1997478#c25

Eviction tests (https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-eviction) are currently broken (https://github.com/kubernetes/kubernetes/issues/102782), there's a PR up to fix: https://github.com/kubernetes/kubernetes/pull/104304

#### Environment:
- Kubernetes version (use `kubectl version`): HEAD/1.23.x (kubeletVersion="v1.23.0-alpha.3.47+d551560a78292e")
- Cloud provider or hardware configuration: local-up-cluster
- OS (e.g: `cat /etc/os-release`), Kernel (e.g. `uname -a`):

```
ehashman@fedora:~/src/k8s$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
ehashman@fedora:~/src/k8s$ uname -a
Linux fedora 5.10.0-8-amd64 #1 SMP Debian 5.10.46-4 (2021-08-03) x86_64 GNU/Linux
```

- Install tools: local-up-cluster.sh
- Network plugin and version (if this is a network-related bug): N/A
- Others: Also reproducible on OpenShift 4.9 (1.23-based)

/sig node
/cc @rphillips @smarterclayton

### PR DESCRIPTION
**Pull request by ehashman**
#### What type of PR is this?

/kind bug
/sig node
/priority critical-urgent

#### What this PR does / why we need it:

Avoids overwriting statuses or phase on terminated pods.

#### Which issue(s) this PR fixes:

Fixes #105358

#### Special notes for your reviewer 

Passing a bool around is ugly, but checking `kl.podWorkers.IsPodKnownTerminated(pod.UID)` is expensive, and we don't need to by the time we invoke `syncTerminatedPod`.

#### Does this PR introduce a user-facing change?

```release-note
Evicted and other terminated pods will no longer revert to Running phase
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ehashman** [comment]
After this patch is applied, on the repro case, I get the following final pod status:

<details>

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"evictme","namespace":"default"},"spec":{"containers":[{"command":["sh","-c","trap 'exit 5' SIGINT SIGTERM; fallocate -l 135G file; sleep 10000"],"image":"k8s.gcr.io/e2e-test-images/nginx:1.14-2","name":"busybox"}],"terminationGracePeriodSeconds":10}}
    creationTimestamp: "2021-10-01T20:01:41Z"
    name: evictme
    namespace: default
    resourceVersion: "399"
    uid: f0d2d9bf-3df7-4840-b30b-03f7f31c8f2b
  spec:
    containers:
    - command:
      - sh
      - -c
      - trap 'exit 5' SIGINT SIGTERM; fallocate -l 135G file; sleep 10000
      image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imagePullPolicy: IfNotPresent
      name: busybox
      resources: {}
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: kube-api-access-dxn45
        readOnly: true
    dnsPolicy: ClusterFirst
    enableServiceLinks: true
    nodeName: 127.0.0.1
    preemptionPolicy: PreemptLowerPriority
    priority: 0
    restartPolicy: Always
    schedulerName: default-scheduler
    securityContext: {}
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 10
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    volumes:
    - name: kube-api-access-dxn45
      projected:
        defaultMode: 420
        sources:
        - serviceAccountToken:
            expirationSeconds: 3607
            path: token
        - configMap:
            items:
            - key: ca.crt
              path: ca.crt
            name: kube-root-ca.crt
        - downwardAPI:
            items:
            - fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
              path: namespace
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:01:48Z"
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:02:06Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:02:06Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:01:48Z"
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: docker://a5c8f64f206a47e2119c945d3865c2deb7316369530b24448ae00991d97931b1
      image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imageID: docker-pullable://k8s.gcr.io/e2e-test-images/nginx@sha256:13616070e3f29de4417eee434a8ef472221c9e51b3d037b5a6b46cef08eb7443
      lastState: {}
      name: busybox
      ready: false
      restartCount: 0
      started: false
      state:
        terminated:
          containerID: docker://a5c8f64f206a47e2119c945d3865c2deb7316369530b24448ae00991d97931b1
          exitCode: 137
          finishedAt: "2021-10-01T20:02:05Z"
          reason: Error
          startedAt: "2021-10-01T20:01:49Z"
    hostIP: 127.0.0.1
    message: 'The node was low on resource: ephemeral-storage. Container busybox was
      using 46866480Ki, which exceeds its request of 0. '
    phase: Failed
    podIP: 172.17.0.2
    podIPs:
    - ip: 172.17.0.2
    qosClass: BestEffort
    reason: Evicted
    startTime: "2021-10-01T20:01:48Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

</details>

Compare to the Running pod output in https://github.com/kubernetes/kubernetes/issues/105358#issue-1011532980

The kubectl client still shows this one as Error rather than Evicted, but that is a client-side bug that we can fix in a follow-up.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1773,7 +1773,8 @@ func (kl *Kubelet) syncTerminatingPod(ctx context.Context, pod *v1.Pod, podStatu
 		return nil
 	}
 
-	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus)
+	// Containers have not yet been killed, so don't mark pod phase as terminated yet
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, false /* not terminated */)
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet.go:1777]: I expected this to be true.  To get to this line, all containers have to be stopped, and no new containers can be started.

**Comment by ehashman** [comment]
This might not be quite enough on its own, if we get a cache miss that pulls in a waiting status; we might need to also apply:

```diff
diff --cc pkg/kubelet/status/status_manager.go
index f61fd526293,f61fd526293..64958b1329b
--- a/pkg/kubelet/status/status_manager.go
+++ b/pkg/kubelet/status/status_manager.go
@@@ -333,7 -333,7 +333,7 @@@ func (m *manager) TerminatePod(pod *v1.
        }
        status := *oldStatus.DeepCopy()
        for i := range status.ContainerStatuses {
--              if status.ContainerStatuses[i].State.Terminated != nil || status.ContainerStatuses[i].State.Waiting != nil {
++              if status.ContainerStatuses[i].State.Terminated != nil {
                        continue
                }
                status.ContainerStatuses[i].State = v1.ContainerState{
@@@ -345,7 -345,7 +345,7 @@@
                }
        }
        for i := range status.InitContainerStatuses {
--              if status.InitContainerStatuses[i].State.Terminated != nil || status.InitContainerStatuses[i].State.Waiting != nil {
++              if status.InitContainerStatuses[i].State.Terminated != nil {
                        continue
                }
                status.InitContainerStatuses[i].State = v1.ContainerState{
```

File: pkg/kubelet/kubelet_pods.go [hunk 2]
```diff
@@ -1433,21 +1433,21 @@ func getPhase(spec *v1.PodSpec, info []v1.ContainerStatus) v1.PodPhase {
...
 	}
-	s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus)
+	s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus, isTerminated)
 
 	// calculate the next phase and preserve reason
 	allStatus := append(append([]v1.ContainerStatus{}, s.ContainerStatuses...), s.InitContainerStatuses...)
 	s.Phase = getPhase(&pod.Spec, allStatus)
 	klog.V(4).InfoS("Got phase for pod", "pod", klog.KObj(pod), "oldPhase", oldPodStatus.Phase, "phase", s.Phase)
-	if s.Phase == oldPodStatus.Phase {
+	if isTerminated || s.Phase == oldPodStatus.Phase {
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_pods.go:1450]: Hrm, terminated transition is important, i don't think a transition TO terminated has the same behavior as is terminated.  Might need a unit test for this where i can review.

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -1469,6 +1469,19 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
 		}
 	}
 
+	// Perform a three-way merge between the statuses from the status manager,
+	// runtime, and generated status. If the pod is terminated, the phase MUST be
+	// terminal. Default to Failed in the case we can't find a non-running Phase.
+	if isTerminated && s.Phase != v1.PodFailed && s.Phase != v1.PodSucceeded {
+		if oldPodStatus.Phase == v1.PodFailed || oldPodStatus.Phase == v1.PodSucceeded {
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_pods.go:1476]: this should be a switch

File: pkg/kubelet/kubelet_pods.go [hunk 4]
```diff
@@ -1469,6 +1469,19 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
...
+			s.Phase = oldPodStatus.Phase
+		} else if pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded {
+			s.Phase = pod.Status.Phase
+		} else {
+			s.Phase = v1.PodFailed
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_pods.go:1481]: Hrm... I don't think we should ever get into this chunk.  Why is this here?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (ehashman) [reply pkg/kubelet/kubelet.go:1777]: this is inside syncTerminatingPod -- you can see that we haven't actually called `killPod` yet, that happens a few lines below: https://github.com/kubernetes/kubernetes/blob/38746752af6d0ec955f4511a727da37347da3b92/pkg/kubelet/kubelet.go#L1788

So I think setting this as true would be premature

File: pkg/kubelet/kubelet_pods.go [hunk 5]
```diff
@@ -1758,9 +1774,11 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			ExitCode: 137,
 		}
 
-		// If the pod was not deleted, then it's been restarted. Increment restart count.
-		if pod.DeletionTimestamp == nil {
-			status.RestartCount += 1
+		if !isTerminated {
+			// If the pod was not deleted, then it's been restarted. Increment restart count.
+			if pod.DeletionTimestamp == nil {
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_pods.go:1779]: Hrm, I don't know that the deletionTimestamp count is relevant?

Comment (smarterclayton) [reply pkg/kubelet/kubelet_pods.go:1779]: I think isTerminated subsumes pod.DeletionTimestamp until we get into syncTerminating, so this if block may not show up.

Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go:1779]: That logic was there previously. This is guarding against the case that we get into this block where we're generating a terminated ContainerStatusUnknown---if we know the pod is terminated, we shouldn't increment the restart count.

File: pkg/kubelet/kubelet_pods.go (hunk 4 above)
Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go:1481]: Is it possible for everything to say Running but we've entered syncTerminatedPod? Trying to be defensive

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/kubelet/kubelet.go:1777]: Oh, haha I was confused by runningPod.  You're absolutely right.

File: pkg/kubelet/kubelet_pods.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/kubelet/kubelet_pods.go:1481]: It's possible we don't have all pod status yet, so all containers will eventually be reported as successful, we just haven't calculated it yet.  If you calculate too early, it'll mark those pods as failed.

File: pkg/kubelet/kubelet_pods.go (hunk 5 above)
Comment (smarterclayton) [reply pkg/kubelet/kubelet_pods.go:1779]: It was there previously as an incorrect substitute for isTerminated.  A pod isn't terminated until the pod worker says it is - so you can have deletion timestamp != nil AND still be running.  For !isTerminated to be true, we have to be in syncPod, and if we're in syncPod, we could still be starting containers, so this execution of syncPod could actually restart a container that when syncPod exits we then terminate.  So this check was historically subtly wrong, and now that we have isTerminated we can reflect that.

**Comment by ehashman** [comment]
Updated patch to hopefully avoid the e2e failure case above (nixed default isTerminated -> Failed).

From what I can see, terminated pods are no longer reporting ContainerCreating, you can see the statuses set on: https://github.com/openshift/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1763-L1767 and https://github.com/kubernetes/kubernetes/blob/38746752af6d0ec955f4511a727da37347da3b92/pkg/kubelet/status/status_manager.go#L339-L344

<details>

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"evictme","namespace":"default"},"spec":{"containers":[{"command":["sh","-c","trap 'exit 5' SIGINT SIGTERM; fallocate -l 135G file; sleep 10000"],"image":"k8s.gcr.io/e2e-test-images/nginx:1.14-2","name":"busybox"}],"terminationGracePeriodSeconds":10}}
    creationTimestamp: "2021-10-01T20:46:23Z"
    name: evictme
    namespace: default
    resourceVersion: "397"
    uid: 4ce46acc-87a0-47d7-91c3-f7f3c05169ff
  spec:
    containers:
    - command:
      - sh
      - -c
      - trap 'exit 5' SIGINT SIGTERM; fallocate -l 135G file; sleep 10000
      image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imagePullPolicy: IfNotPresent
      name: busybox
      resources: {}
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: kube-api-access-bg4gb
        readOnly: true
    dnsPolicy: ClusterFirst
    enableServiceLinks: true
    nodeName: 127.0.0.1
    preemptionPolicy: PreemptLowerPriority
    priority: 0
    restartPolicy: Always
    schedulerName: default-scheduler
    securityContext: {}
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 10
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    volumes:
    - name: kube-api-access-bg4gb
      projected:
        defaultMode: 420
        sources:
        - serviceAccountToken:
            expirationSeconds: 3607
            path: token
        - configMap:
            items:
            - key: ca.crt
              path: ca.crt
            name: kube-root-ca.crt
        - downwardAPI:
            items:
            - fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
              path: namespace
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:46:25Z"
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:46:43Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:46:43Z"
      message: 'containers with unready status: [busybox]'
      reason: ContainersNotReady
      status: "False"
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: "2021-10-01T20:46:25Z"
      status: "True"
      type: PodScheduled
    containerStatuses:
    - image: k8s.gcr.io/e2e-test-images/nginx:1.14-2
      imageID: ""
      lastState:
        terminated:
          exitCode: 137
          finishedAt: null
          message: The container could not be located when the pod was deleted.  The
            container used to be Running
          reason: ContainerStatusUnknown
          startedAt: null
      name: busybox
      ready: false
      restartCount: 0
      started: false
      state:
        terminated:
          exitCode: 137
          finishedAt: null
          message: The container could not be located when the pod was terminated
          reason: ContainerStatusUnknown
          startedAt: null
    hostIP: 127.0.0.1
    message: 'The node was low on resource: ephemeral-storage. Container busybox was
      using 14319660Ki, which exceeds its request of 0. '
    phase: Failed
    podIP: 172.17.0.2
    podIPs:
    - ip: 172.17.0.2
    qosClass: BestEffort
    reason: Evicted
    startTime: "2021-10-01T20:46:25Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
</details>

File: pkg/kubelet/kubelet_pods.go (hunk 2 above)
Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go:1450]: Another option here would be to move the phase merge above this block which will force set `s.Phase == oldPodStatus.Phase` in the case of Failed/Succeeded

File: pkg/kubelet/kubelet_pods.go (hunk 5 above)
Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go]: Looks like if we change this we have unit test failures... I guess we should decide which behaviour we want.

Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go]: Made the change suggested in https://github.com/kubernetes/kubernetes/pull/105411#discussion_r720546489 and updated the unit tests, split that into a separate commit so it can be dropped.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/kubelet/kubelet.go:1777]: The inline quote is not necessary now that you have the previous line

File: pkg/kubelet/kubelet_test.go [hunk 6]
```diff
@@ -2243,7 +2243,7 @@ func TestGenerateAPIPodStatusInvokesPodSyncHandlers(t *testing.T) {
 		Name:      pod.Name,
 		Namespace: pod.Namespace,
 	}
-	apiStatus := kubelet.generateAPIPodStatus(pod, status)
+	apiStatus := kubelet.generateAPIPodStatus(pod, status, false)
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_test.go:2246]: We probably want to have a unit test where we pass true.

Also a unit test for the scenario we detected here (verifying a transition when runtime cache status is running, status manager has terminal phase failed for eviction, and the passed in pod is in Running state).

File: pkg/kubelet/status/status_manager.go [hunk 7]
```diff
@@ -334,7 +334,7 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 	}
 	status := *oldStatus.DeepCopy()
 	for i := range status.ContainerStatuses {
-		if status.ContainerStatuses[i].State.Terminated != nil || status.ContainerStatuses[i].State.Waiting != nil {
+		if status.ContainerStatuses[i].State.Terminated != nil {
```
Comment (smarterclayton) [inline pkg/kubelet/status/status_manager.go:337]: Probably need to add a new TestTerminatePodWithWaiting

File: pkg/kubelet/kubelet_pods.go [hunk 8]
```diff
@@ -1433,20 +1433,34 @@ func getPhase(spec *v1.PodSpec, info []v1.ContainerStatus) v1.PodPhase {
...
+	s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus, isTerminated)
 
 	// calculate the next phase and preserve reason
 	allStatus := append(append([]v1.ContainerStatus{}, s.ContainerStatuses...), s.InitContainerStatuses...)
 	s.Phase = getPhase(&pod.Spec, allStatus)
 	klog.V(4).InfoS("Got phase for pod", "pod", klog.KObj(pod), "oldPhase", oldPodStatus.Phase, "phase", s.Phase)
+
+	// Perform a three-way merge between the statuses from the status manager,
+	// runtime, and generated status to ensure terminal status is correctly set.
+	if isTerminated && s.Phase != v1.PodFailed && s.Phase != v1.PodSucceeded {
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet_pods.go:1453]: Thinking through a possible scenario:

1. Pod is running on both apiserver and in kubelet
2. Another client sets pod phase = Failed via apiserver
3. Config update sees it, passes it to pod worker, pod worker sets becomeTerminating = true
4. Pod worker loop enters syncTerminatingPod, calls this method with `false`
5. This loop is bypassed, s.phase is Running
6. s.phase Running is passed to pod sync handler (which is wrong)
7. pod.status.phase is Running, so we don't trigger illegal phase transition

So we're missing an condition on this check.  We have to explicitly check all three sources for pod terminal, and then preserve that.

The guarantee this check has to have is if any source (s.phase, pod.Status.Phase, or oldPodStatus.Phase) is terminal, s.Phase has to be terminal as well.  The current chunk does not guarantee that.

Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go:1453]: s is calculated from `s := kl.convertStatusToAPIStatus(pod, podStatus, oldPodStatus, isTerminated)`. Is it possible for `s` to ever be terminal if either `pod` or `oldPodStatus` is non-terminal?

Comment (smarterclayton) [reply pkg/kubelet/kubelet_pods.go:1453]: convertStatusToAPIStatus ignores phase, and isTerminated would be false, so in the scenario I expected in 4 that s.phase is Running (but correct me if I'm wrong)

Comment (ehashman) [reply pkg/kubelet/kubelet_pods.go:1453]: > This loop is bypassed because s.phase is Running

This block is bypassed because `isTerminated` is false, do we just want to drop that requirement? I can test that.

**Comment by ehashman** [comment]
Closing in favour of https://github.com/kubernetes/kubernetes/pull/105462
